pycsw/wsgi.py

```diff
@@ -55,18 +30,53 @@
 import os
 import sys
 import six
+import re
+import ConfigParser
 from six.moves.urllib.parse import unquote
 
 from pycsw import server
 
 
 PYCSW_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
 
+def dynUpdateConfigurationURLs(new_url):
+    '''
+    Update dynamically PyCSW configuration to match public IP:port of this server.
+
+    The PyCSW server can be deployed in a VM or a container in different environment (dev, prod...). Its IP:port is therefore not deterministic.
+    The values of 'serveraddress' in default.cfg can't be forecasted in advance.
+
+    This function is meant to be called the client emit a request.
+    The base URL requested by the client is retrieved and reinjected in PyCSW configuration.
+    '''
+    confFile = os.environ.get('PYCSW_CONFIG', '/etc/pycsw/default.cfg')
+    configpar = ConfigParser.ConfigParser()
+    configpar.read(confFile)
+    # Old values
+    oldServerURL = configpar.get('server', 'url')
+
+    # New values
+    new_url = re.sub('^https?://', '', new_url)  # Remove http[s] if present
+    newServerURL = 'http://' + new_url + '/pycsw/csw.py'
+
+    # Update the config is URLs are not the same as the one used for this HTTP requet.
+    if oldServerURL != newServerURL:
```

Comment:
I wonder what is the point in keeping the `server.url` parameter in the configuration at all. Since the point of this PR is to make this parameter become dynamic then it seems that it should be removed from the config altogether and just used in the bits of the code that require it (mostly in the GetCapabilities operation I guess). This would negate the bug mentioned in the previous comment and actually seems like a better implementation IMHO.
It is also misleading and magical to a catalogue administrator to have a configuration parameter be changed at runtime.